src/main/java/org/apache/maven/plugin/compiler/TestCompilerMojo.java

```diff
@@ -319,7 +352,42 @@ else if ( Double.valueOf( getTarget() ) < Double.valueOf( MODULE_INFO_TARGET ) )
         if ( testModuleDescriptor != null )
         {
             modulepathElements = testPath;
-            classpathElements = Collections.emptyList();
+            classpathElements = null;
+            for ( String path : modulepathElements )
+            {
+                File file = new File( path );
+                if ( file.exists() && ! file.isDirectory() && file.getName().endsWith( ".jar" ) )
+                {
+                    // check if the jar file contains the module-info.class
+                    try
+                    {
+                        JarFile jarFile = new JarFile( file );
+                        if ( jarFile.getJarEntry( "module-info.class" ) == null )
+                        {
+                            if ( getLog().isDebugEnabled() )
+                            {
+                                getLog().debug( "Add non-module jar dependency " + path
+                                        + " to class path." );
+                            }
+                            if ( classpathElements ==  null )
+                            {
+                                classpathElements = new ArrayList<String>();
+                            }
+                            // add the path to the classpathElements list.
+                            classpathElements.add( path );
+                        }
```

Comment:
Can you refactor the code a bit? Too many levels of indentation.

Reply:
Not sure exactly how you want me to refactor this code.  The build checkstyle definition forced me into this code style.  I did combine the debug statement within a single line.  The only other option is to extract this set of code and create a private method.  That would reduce the indentation.  Is this what you were thinking?

Reply:
Yes that is what I had in mind. A small comment why stuff happens there would be helpful as well. I would start a little above in the "try" statement with the comment: `                    // check if the jar file contains the module-info.class.`.
I think the whole try statement would make a nice method. A simple comment why the jars are explicitly added to the classpath would absolutely suffice.

Also: 
`new ArrayList<String>();` -- you can just use the diamond operator: `new ArrayList<>();`. We are on Java 7, not 5. ;-) 

Thanks!

Reply:
I went a step further, I ended up extracting the for loop into a separate method.

Reply:
Well, now you have the same amount of indentation in a loop. But I will not complain anymore about indentation unless another committer does. ;-)
The comment in the new method is helpful. Nit: You could have added it as Javadoc while you were at it.

I will check it out locally in a few days to inspect the changes further.